### PR TITLE
fix: stop MTP attack

### PR DESCRIPTION
**Pull request by SWvheerden**
Description
---
This changes the timestamp of a new block template to be always greater than the median time past (MTP) of the past blocks as per consensus. 

Motivation and Context
---
It is possible for an attack to forward the median time to be greater than now, but less than the future time limit(FTL). This means that all valid blocks with the current timestamp will be rejected. This will ensure that ll newly created blocks always have a timestamp greater than the MTP. Due to consensus we also know that MTP will always be less than FTL. This ensures that all new block timestamps are `MTP < Timestamp < FTL`

For a description see: [Graft network](https://github.com/graft-project/GraftNetwork/pull/118#issuecomment-384599435)
and [MTP Attack (Jagernan)](https://github.com/zawy12/difficulty-algorithms/issues/30)

How Has This Been Tested?
---

All current unit tests pass.

### CODE REVIEW

File: applications/tari_mining_node/src/difficulty.rs [hunk 1]
```diff
@@ -76,7 +76,11 @@ impl BlockHeaderSha3 {
     }
 
     pub fn set_timestamp(&mut self, timestamp: u64) {
-        self.timestamp = timestamp;
+        // if the timestamp has been advanced by the base_node due to the median time we should not reverse it but we
+        // should only change the timestamp if we move it forward.
+        if timestamp > self.timestamp {
```
Comment (sdbondi) [inline applications/tari_mining_node/src/difficulty.rs:83]: nit: I'd rather go for putting the check outside the function call (it's only used in one place anyhow). It isn't expected for `set_timestamp` not to set the timestamp. Alternatively, name the function `set_forward_timestamp` (or something) and explain in a docstring that is only sets the timestamp forward from what it is. This is a nit so feel free to ignore.

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 2]
```diff
@@ -643,7 +650,24 @@ where B: BlockchainBackend
     pub fn prepare_block_merkle_roots(&self, template: NewBlockTemplate) -> Result<Block, ChainStorageError> {
         let NewBlockTemplate { header, mut body, .. } = template;
         body.sort();
-        let header = BlockHeader::from(header);
+        let mut header = BlockHeader::from(header);
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs:653]: This method now does more than it says it does. It should be renamed or this logic should be moved elsewhere

Comment (sdbondi) [inline base_layer/core/src/chain_storage/blockchain_database.rs:653]: nit: We may as well call this function `prepare_new_block(NewBlocktemplate) -> Block`

**Review by sdbondi — DISMISSED** [review]
Nice catch - some nits that you can safely ignore :)

File: applications/tari_mining_node/src/difficulty.rs (hunk 1 above)
Comment (SWvheerden) [reply applications/tari_mining_node/src/difficulty.rs:83]: I agree good idea

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 2 above)
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/blockchain_database.rs:653]: I agree, will rename. That is what the function did anyway. All we do now is we do an extra check before we set the timestamp. Before this, the set_timestamp happened in `let header = BlockHeader::from(header);`

Comment (SWvheerden) [reply base_layer/core/src/chain_storage/blockchain_database.rs:653]: I renamed the method to `prepare_new_block`. The timestamp was already set by this method, it was done inside of `let header = BlockHeader::from(header);`. All it does not is add an extra check to what it should set it and not just blanketly set it to now. But I do agree with the fact that the function name and what it does is misaligned.

**Review by stringhandler — APPROVED** [review]

**Pull request merged**